pkg/cosign/verify.go

```diff
@@ -291,6 +291,88 @@ func ValidateAndUnpackCert(cert *x509.Certificate, co *CheckOpts) (signature.Ver
 	return verifier, nil
 }
 
+// ValidateAndUnpackCertWithIntermediates function has same logic as ValidateAndUnpackCert.
+// Main diffrence is that we can not use ValidateAndUnpackCert in multi thread invocation
+// as it uses a common reference of CheckOpts so in multi thread invocation co.intermediateCert object
+// can get overwritten my multiple threads. So to solve this ValidateAndUnpackCertWithIntermediates
+// is created which has a separate argument for Intermediate certs.
+func ValidateAndUnpackCertWithIntermediates(cert *x509.Certificate, co *CheckOpts, intermediateCerts *x509.CertPool) (signature.Verifier, error) {
+	verifier, err := signature.LoadVerifier(cert.PublicKey, crypto.SHA256)
+	if err != nil {
+		return nil, fmt.Errorf("invalid certificate found on signature: %w", err)
+	}
+
+	// Handle certificates where the Subject Alternative Name is not set to a supported
+	// GeneralName (RFC 5280 4.2.1.6). Go only supports DNS, IP addresses, email addresses,
+	// or URIs as SANs. Fulcio can issue a certificate with an OtherName GeneralName, so
+	// remove the unhandled critical SAN extension before verifying.
+	if len(cert.UnhandledCriticalExtensions) > 0 {
+		var unhandledExts []asn1.ObjectIdentifier
+		for _, oid := range cert.UnhandledCriticalExtensions {
+			if !oid.Equal(cryptoutils.SANOID) {
+				unhandledExts = append(unhandledExts, oid)
+			}
+		}
+		cert.UnhandledCriticalExtensions = unhandledExts
+	}
+
+	// Now verify the cert, then the signature.
+	if intermediateCerts == nil {
```

Comment:
I would remove this conditional, as the purpose of this function is to pass in the intermediates that should be used for chain building.

Reply:
This check is important and should be present.
we will reach the new function ValidateAndUnpackCertWithIntermediates() via two paths
1. ValidateAndUnpackCert()
2. verifyInternal()

verifyInternal is called in some cases where we dont have an intermediate present in signature but we pass that as an argument in verification in those cases co.IntermediateCerts will have a valid intermediate cert but we will not be able to set the intermediatePool argument. one scenario is verification of Blob signatures.
so we need to have this condition if the intermediate Argument is Null that set the value from  co.IntermediateCerts
I tested this by removing this condition and i could see verify blob signature test cases were failing so i put this condition back.

Reply:
Ah I think I understand the issue. Correct me if I misunderstand, you're saying that co.IntermediateCerts may have been set by one of the functions that calls verifyInternal.

I think the fix would just be to pass `co.IntermediateCerts` to `ValidateAndUnpackCertWithIntermediates` on line 745 if `co.IntermediateCerts != nil`. Then you could remove this check.